llvm/lib/CASObjectFormats/SimpleV1.cpp

```diff
@@ -0,0 +1,1217 @@
+  if (!E)
+    E = encoding::consumeVBR8(Remaining, NormalizedTripleSize);
+  if (!E)
+    E = encoding::consumeVBR8(Remaining, Endianness);
+  if (E) {
+    consumeError(std::move(E));
+    return createStringError(inconvertibleErrorCode(),
+                             "corrupt compile unit data");
+  }
+  if (Endianness != uint8_t(support::endianness::little) &&
+      Endianness != uint8_t(support::endianness::big))
+    return createStringError(inconvertibleErrorCode(),
+                             "corrupt compile unit endianness");
+
+  auto TripleStr = consumeDataOfSize(Remaining, NormalizedTripleSize);
+  if (!TripleStr)
+    return TripleStr.takeError();
+
+  return CompileUnitRef(*Specific);
+}
+
+Expected<CompileUnitRef> CompileUnitRef::create(const ObjectFileSchema &Schema,
+                                                const jitlink::LinkGraph &G,
+                                                raw_ostream *DebugOS) {
+  if (Error E = helpers::checkArch(G.getTargetTriple()))
+    return std::move(E);
+
+  CompileUnitBuilder Builder(Schema, G.getTargetTriple(), DebugOS);
+
+  // Visit sections first since they don't have references to each other.
+  // Assume section list has a stable ordering.
+  for (const auto &S : G.sections()) {
+    auto Ref = Builder.createSection(S);
+    if (!Ref)
+      return Ref.takeError();
+    Builder.commitNode(*Ref);
+  }
+  // Visit symbols. Sort the symbols for stable ordering.
+  // FIXME: duplicated code for sorting and comparsion.
+  SmallVector<const jitlink::Symbol *, 16> Symbols;
+  auto appendSymbols = [&](auto &&NewSymbols) {
+    size_t PreviousSize = Symbols.size();
+    Symbols.append(NewSymbols.begin(), NewSymbols.end());
+    llvm::sort(Symbols.begin() + PreviousSize, Symbols.end(),
+               compareSymbolsByAddress);
+  };
+  for (const jitlink::Section &Section : G.sections())
+    appendSymbols(Section.symbols());
+  appendSymbols(G.absolute_symbols());
+  appendSymbols(G.external_symbols());
+
+  std::stable_sort(Symbols.begin(), Symbols.end(),
+                   compareSymbolsByLinkageAndSemantics);
+
+  // Visit blocks. Create a ordered list of blocks so it can be index.
+  auto appendBlocks = [&](auto &&NewBlocks) {
+    size_t PreviousSize = Builder.Blocks.size();
+    Builder.Blocks.append(NewBlocks.begin(), NewBlocks.end());
+    llvm::sort(Builder.Blocks.begin() + PreviousSize, Builder.Blocks.end(),
+               compareBlocksByAddress);
+  };
+  for (const jitlink::Section &Section : G.sections())
+    appendBlocks(Section.blocks());
+
+  // Create sections.
+  for (auto *S : Symbols) {
+    auto Ref = Builder.createSymbol(*S);
+    if (!Ref)
+      return Ref.takeError();
+    Builder.commitNode(*Ref);
+  }
+
+  // Create BlockRefs.
+  Builder.BlockIndexStarts.reserve(Builder.Blocks.size());
+  for (auto *B : Builder.Blocks) {
+    auto Ref = Builder.createBlock(*B);
+    if (!Ref)
+      return Ref.takeError();
+    Builder.commitNode(*Ref);
+  }
```

Comment:
I wonder if it'd be more convenient to create blocks first, then symbols, since at link graph creation time the simple order is:
- create blocks
- create symbols
- add edges

Reply:
So current the order is a bit different for `create` and `materialize`. `materialize` follow the same order as you described here to go from CASObject -> LinkGraph. LinkGraph -> CASObject actually does symbols then blocks as it can encode edges going through blocks (which references symbol).